File: src/main/java/xyz/nkomarn/composter/network/protocol/packet/play/ClientboundChunkOperationPacket.java
package xyz.nkomarn.composter.network.protocol.packet.play;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import org.jetbrains.annotations.NotNull;
import xyz.nkomarn.composter.network.protocol.Packet;

public class ClientboundChunkOperationPacket extends Packet<ClientboundChunkOperationPacket> {

    private int x;
    private int z;
    private boolean mode;

    public ClientboundChunkOperationPacket() {
    }

    public ClientboundChunkOperationPacket(int x, int z, boolean mode) {
        this.x = x;
        this.z = z;
        this.mode = mode;
    }

    @Override
    public int getId() {
        return 0x32;
    }

    @Override
    public @NotNull ByteBuf encode() {
        return Unpooled.buffer()
                .writeInt(x)
                .writeInt(z)
                .writeBoolean(mode);
    }
}

File: src/main/kotlin/kyta/composter/protocol/packet/play/ClientboundChunkOperationPacket.kt
package kyta.composter.protocol.packet.play

import kyta.composter.protocol.Packet
import kyta.composter.protocol.PacketSerializer
import kyta.composter.protocol.WriteBuffer
import kyta.composter.world.ChunkPos

data class ClientboundChunkOperationPacket(val pos: ChunkPos, val mode: Mode) : Packet {
    enum class Mode {
        LOAD,
        UNLOAD,
    }

    companion object : PacketSerializer<ClientboundChunkOperationPacket> {
        override fun serialize(packet: ClientboundChunkOperationPacket, buffer: WriteBuffer) {
            buffer.writeInt(packet.pos.x)
            buffer.writeInt(packet.pos.z)
            buffer.writeBoolean(packet.mode == Mode.LOAD)
        }
    }
}
